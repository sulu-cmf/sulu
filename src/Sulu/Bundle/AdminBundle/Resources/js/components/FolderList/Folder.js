// @flow
import React from 'react';
import Icon from '../Icon';
import folderStyles from './folder.scss';

const FOLDER_ICON = 'folder-o';

type Props = {
    /** The id which will be passed as an argument inside the onClick handler */
    id: string | number,
    /** The subtext underneath the title */
    info: string,
    title: string,
    onClick?: (id: string | number) => void,
};

export default class Folder extends React.PureComponent<Props> {
    handleClick = () => {
        if (this.props.onClick) {
            this.props.onClick(this.props.id);
        }
    };

    render() {
        const {
            info,
            title,
        } = this.props;

        return (
            <div
                className={folderStyles.folder}
                onClick={this.handleClick}
                role="button"
                tabIndex="0"
            >
                <div className={folderStyles.iconContainer}>
                    <Icon name={FOLDER_ICON} />
                </div>
                <div className={folderStyles.description}>
                    <h5 className={folderStyles.title}>
                        {title}
                    </h5>
                    <div className={folderStyles.info}>
                        {info}
                    </div>
                </div>
            </div>
        );
    }
}
