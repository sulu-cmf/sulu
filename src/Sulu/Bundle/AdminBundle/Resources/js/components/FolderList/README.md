The `FolderList` component is used to display a list of folders. The `FolderList` accepts `Folder` components as 
children. The `Folder` component is just an interactive element for presentation purposes. Every `Folder` component has 
its own `onClick`-handler.

```
const Folder = require('./Folder').default;

const handleClick = (id) => {
    alert(`You clicked on a Folder, wohooo, do you want a cookie now?`)
};

<Folder
    id="1"
    info="3 Objects"
    title="This is a folder"
    onClick={handleClick}
/>
```

The `FolderList` component offers a more convenient way to work with multiple folders by offering 
a single `onClick`-handler:

```
const handleClick = (id) => {
    alert(`Folder with the id "${id}" was clicked`)
};

<FolderList onFolderClick={handleClick}>
    <FolderList.Folder
        id="1"
        info="3 Objects"
        title="This is a folder"
    />
    <FolderList.Folder
        id="2"
        info="0 Objects"
        title="I am empty"
    />
    <FolderList.Folder
        id="3"
        info="1 Object"
        title="I have 1 child inside me"
    />
    <FolderList.Folder
        id="4"
        info="10 Objects"
        title="And I have 10 children. Wuhuu!"
    />
    <FolderList.Folder
        id="5"
        info="2 Objects"
        title="Jeez, these folders..."
    />
    <FolderList.Folder
        id="6"
        info="4 Objects"
        title="el oh el"
    />
</FolderList>
```
