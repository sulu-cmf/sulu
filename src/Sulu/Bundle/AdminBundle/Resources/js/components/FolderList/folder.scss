@import '../../containers/Application/colors.scss';

$folderBackgroundColor: $silver;
$metaFontColor: $silverChalice;

.folder {
    width: 240px;
    height: 80px;
    display: flex;
    padding: 10px;
    border: 1px solid $folderBackgroundColor;
    border-radius: 3px;
    cursor: pointer;
}

.description {
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-left: 10px;
}

.title {
    overflow: hidden;
    max-height: 35px;
    font-size: 12px;
    line-height: 18px;
    margin: 0 0 5px;
}

.info {
    font-size: 10px;
    color: $metaFontColor;
}

.icon-container {
    padding: 10px;
    font-size: 26px;
}
