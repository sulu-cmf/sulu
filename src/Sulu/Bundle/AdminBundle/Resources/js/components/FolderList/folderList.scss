.folder-list {
    margin: 0;
    padding: 0;
    list-style: none;

    & > li {
        display: inline-block;
        margin: 0 30px 30px 0;
    }
}
