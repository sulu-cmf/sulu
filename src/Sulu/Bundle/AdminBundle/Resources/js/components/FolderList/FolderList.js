// @flow
import React from 'react';
import type {ChildrenArray, Element} from 'react';
import Folder from './Folder';
import folderListStyles from './folderList.scss';

type Props = {
    children: ChildrenArray<Element<typeof Folder>>,
    onFolderClick?: (folderId: string | number) => void,
};

export default class FolderList extends React.PureComponent<Props> {
    static Folder = Folder;

    cloneFolders(originalFolders: ChildrenArray<Element<typeof Folder>>) {
        return React.Children.map(originalFolders, (folder) => (
            <li>
                {
                    React.cloneElement(
                        folder,
                        {
                            ...folder.props,
                            onClick: this.handleFolderClick,
                        },
                    )
                }
            </li>
        ));
    }

    handleFolderClick = (folderId: string | number) => {
        if (this.props.onFolderClick) {
            this.props.onFolderClick(folderId);
        }
    };

    render() {
        const {children} = this.props;
        const clonedFolders = this.cloneFolders(children);

        return (
            <ul className={folderListStyles.folderList}>
                {clonedFolders}
            </ul>
        );
    }
}
