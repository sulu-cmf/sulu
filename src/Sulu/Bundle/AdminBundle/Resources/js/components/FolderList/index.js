// @flow
import FolderList from './FolderList';
import Folder from './Folder';

export default FolderList;
export {Folder};
