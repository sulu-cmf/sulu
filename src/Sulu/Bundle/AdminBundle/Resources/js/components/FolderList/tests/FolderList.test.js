/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import {render, mount} from 'enzyme';
import FolderList from '../FolderList';

test('Render an empty FolderList', () => {
    expect(render(<FolderList />)).toMatchSnapshot();
});

test('Render a FolderList with Folder components inside', () => {
    expect(render(
        <FolderList>
            <FolderList.Folder
                id="1"
                info="3 Objects"
                title="This is a folder"
            />
            <FolderList.Folder
                id="2"
                info="2 Objects"
                title="This is a folder"
            />
            <FolderList.Folder
                id="3"
                info="0 Objects"
                title="This is a folder"
            />
        </FolderList>
    )).toMatchSnapshot();
});

test('Clicking on a folder should call the click handler with the right id as argument', () => {
    const clickSpy = jest.fn();
    const clickedFolderId = 3;
    const folderList = mount(
        <FolderList onFolderClick={clickSpy}>
            <FolderList.Folder
                id="1"
                info="3 Objects"
                title="This is a folder"
            />
            <FolderList.Folder
                id="2"
                info="2 Objects"
                title="This is a folder"
            />
            <FolderList.Folder
                id={clickedFolderId}
                info="0 Objects"
                title="This is a folder"
            />
        </FolderList>
    );

    folderList.find('.folder').at(2).simulate('click');
    expect(clickSpy).toHaveBeenCalledWith(clickedFolderId);
});
