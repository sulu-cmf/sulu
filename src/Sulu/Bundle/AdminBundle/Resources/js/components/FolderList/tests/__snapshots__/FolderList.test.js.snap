// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a FolderList with Folder components inside 1`] = `
<ul
  class="folderList"
>
  <li>
    <div
      class="folder"
      role="button"
      tabindex="0"
    >
      <div
        class="iconContainer"
      >
        <span
          aria-hidden="true"
          class="fa fa-folder-o"
        />
      </div>
      <div
        class="description"
      >
        <h5
          class="title"
        >
          This is a folder
        </h5>
        <div
          class="info"
        >
          3 Objects
        </div>
      </div>
    </div>
  </li>
  <li>
    <div
      class="folder"
      role="button"
      tabindex="0"
    >
      <div
        class="iconContainer"
      >
        <span
          aria-hidden="true"
          class="fa fa-folder-o"
        />
      </div>
      <div
        class="description"
      >
        <h5
          class="title"
        >
          This is a folder
        </h5>
        <div
          class="info"
        >
          2 Objects
        </div>
      </div>
    </div>
  </li>
  <li>
    <div
      class="folder"
      role="button"
      tabindex="0"
    >
      <div
        class="iconContainer"
      >
        <span
          aria-hidden="true"
          class="fa fa-folder-o"
        />
      </div>
      <div
        class="description"
      >
        <h5
          class="title"
        >
          This is a folder
        </h5>
        <div
          class="info"
        >
          0 Objects
        </div>
      </div>
    </div>
  </li>
</ul>
`;

exports[`Render an empty FolderList 1`] = `
<ul
  class="folderList"
/>
`;
